tests/testthat/test-bind.R

```diff
@@ -1093,6 +1093,34 @@ test_that("vec_rbind() only restores one time", {
   ))
 })
 
+test_that("vec_rbind() only restores one time (data frame case)", {
+  restored <- list()
+
+  local_methods(
+    vec_ptype2.vctrs_foobar.vctrs_foobar = function(x, y, ...) foobar(df_ptype2(x, y, ...)),
+    vec_cast.vctrs_foobar.vctrs_foobar = function(x, to, ...) foobar(df_cast(x, to, ...)),
+    vec_proxy.vctrs_foobar = function(x, ...) x,
```

Comment:
So I guess the alternative to this PR is to require that data frame subclasses provide a `vec_proxy()` method that just strips the class, which I then realized is what you advised here https://github.com/r-lib/vctrs/issues/1129#issuecomment-635991682

I agree that this PR is a good improvement, even if a little awkward